quodlibet/quodlibet/formats/_audio.py

```diff
@@ -536,24 +537,77 @@ def _role_call(self, role_tag, sub_keys=None):
-            filename = escape_filename(filename)
-            sub_dir = escape_filename(sub_dir)
-        else:
-            filename = fsnative(filename)
-            sub_dir = fsnative(sub_dir)
-
-        path = os.path.join(
-            expanduser(fsnative(u"~/.lyrics")), sub_dir, filename)
-        return path
+        """Returns the validated, or default, lyrics filename for this
+        file. User defined '[memory] lyric_rootpaths' and
+        '[memory] lyric_filenames' matches take precedent"""
+
+        from quodlibet.pattern import ArbitraryExtensionFileFromPattern
+
+        lyric_paths = \
+            config.getstringlist("memory", "lyric_rootpaths", [])
+        # add default
+        lyric_paths.append(u'~/.lyrics')
+        lyric_filenames = \
+            config.getstringlist("memory", "lyric_filenames", [])
+        # add defaults
+        lyric_filenames.append(
+            (self.comma("lyricist") or self.comma("artist"))
+            .replace(u'/', u'')[:128] + os.sep +
+            self.comma("title").replace(u'/', u'')[:128] + u'.lyric')
+        lyric_filenames.append(
+            (self.comma("lyricist") or self.comma("artist"))
+            .replace(u'/', u'')[:128] + ' - ' +
+            self.comma("title").replace(u'/', u'')[:128] + u'.lyric')
+
+        # generate all paths
+        pathfiles = OrderedDict()
+        pathfiles_expanded = OrderedDict()
+        for p in lyric_paths:
+            for f in lyric_filenames:
+                pathfile = ""
+                if os.name == "nt":
+                    # this was added at a later point. only use
+                    # escape_filename here to keep the linux case
+                    # the same as before
+                    pathfile = \
+                        os.path.join(p, os.path.dirname(f),
+                                     escape_filename(os.path.basename(f)))
+                else:
+                    pathfile = \
+                        os.path.join(p, os.path.dirname(f),
+                                    fsnative(os.path.basename(f)))
+                if not pathfile in pathfiles:
+                    pathfiles[pathfile] = pathfile
+
+        #print_d("looking for lyrics files:\n%s" % '\n'.join(pathfiles.keys()))
+
+        match_ = ""
+        for pathfile in pathfiles.keys():
+            path = expanduser(pathfile)
+            if '<' in path:
+                path = ArbitraryExtensionFileFromPattern(pathfile).format(self)
+            pathfiles_expanded[path] = path  # resolved as late as possible
+            if os.path.exists(path):
+                match_ = path
+                break
+
+        if not match_:
+            # look at modified extensions
+            lyric_extensions = ('lyric', 'lyrics', '', 'txt')
+            for pathfile in pathfiles_expanded.keys():
+                ext = os.path.splitext(pathfile)[1][1:]
+                path = pathfile[:-1 * len(ext)].strip('.') if ext else pathfile
```

Comment:
I'm having trouble understanding what this logic is trying to do - it's fairly highly nested with lots of paths through the logic. Can it be simplified / methods extracted to make it clearer?

Reply:
good call again, i've blitzed in with comments and pulled out a big chunk of the logic to  the 'generate_mod_ext_paths' nested function whose name denotes purpose